R/Quosure.R

```diff
@@ -0,0 +1,270 @@
+  get(var, envir = object@env)
+})
+
+#' @param x (`Quosure`)
+#' @param i (`character`) name of the binding in environment (name of the variable)
+#' @param j not used
+#' @param ... not used
+#' @rdname get_var
+#' @export
+setMethod("[[", c("Quosure", "ANY", "missing"), function(x, i, j, ...) {
+  get_var(x, i)
+})
+
+
+#' Get the code from the `Quosure`
+#'
+#' @param object (`Quosure`)
+setMethod("get_code", signature("Quosure"), function(object) {
+  object@code
+})
+
+#' Join two `Quosure` objects
+#'
+#' Combine two `Quosure` object by merging their environments and the code.
+#' Not every `Quosure` objects can be combined:
+#' - if their environments contains objects of the same name but not identical
+#' - if `object2` has unique code element placed before common element. This means that `object2`
+#' in the environment of the `object2` was evaluated some extra code before which can influence
+#' reproducibility
+#' - more cases todo
+#' @param object (`Quosure`)
+#' @param object2 (`Quosure`)
+#' @param overwrite (`logical(1)`) whether modified objects in the environment of the `object2`
+#'  should overwrite their equivalents in the environment of the `object`.
+#' @export
+setGeneric("join", function(object, object2, overwrite = FALSE) {
+  standardGeneric("join")
+})
+
+setMethod("join", signature("Quosure", "Quosure"), function(object, object2, overwrite = FALSE) {
+  # todo: revise chunks_merge_chunks and make robust join method
+
+  # object2 can't have modified object of the same name! See chunks_push_chunks
+  common_names <- intersect(ls(object@env), ls(object2@env))
+  is_identical_obj <- vapply(common_names, function(x) {
+    identical(get(x, object@env), get(x, object2@env))
+  }, logical(1))
+  if (any(isFALSE(is_identical_obj)) && !overwrite) {
+    stop(
+      "join does not allow overwriting already calculated values.",
+      " Following variables would have been overwritten:",
+      paste("    -", names(is_identical_obj), collapse = "\n"),
+      sep = "\n"
+    )
+  }
+
+  # join expressions
+  # > join should be an inteligent union
+  unique_expr <- !(object2@code %in% object@code & names(object2@code) %in% names(object@code))
+    # duplicated expr will probably have the same name also
+    # unique_expr should have TRUE only at the end of the vector.
+  object@code <- c(object@code, object2@code[unique_expr])
+
+  # insert new objects to env
+  new_names <- setdiff(ls(object2@env), ls(object@env))
+
+  object@env <- .copy_env(object@env)
+  lapply(new_names, function(x) assign(x, get(x, object2@env), object@env))
+
+  object
+})
+
+#' Copy environment content to the other environment
+#' To avoid sharing the same environment by multiplequo object (in reactives)
+#' @keywords internal
+.copy_env <- function(env) {
+  # need to copy the objects from old env to new env
+  # to avoid "chunks" sharing the same environment
+  env_new <- new.env()
+  for (i in ls(env)) {
+    env_new[[i]] <- env[[i]]
```

Comment:
- Could this be really slow? What happens if there's lots of data here?  Could you handle this by having the parent of the new environment be say the old ones? Though I guess that could get more complex. I could also imagine other complex things that could but probably shouldn't be done here. 

- Do you actually want separate environments for everything - could you have a separate single "evaluation environment" for the shiny module? Would that be easier for module developers to grasp or harder?

- Also what happens if the parent environments of the two objects don't match - you could get some weird behaviour joining them in that case? 

I guess I'm trying to get as intuitive (if you already know shiny) behaviour as possible here - could the output of the quosure actually be available in the module's environment? So that they behave as similar to reactives as possible.

Reply:
> Could this be really slow? What happens if there's lots of data here? Could you handle this by having the parent of the new environment be say the old ones? Though I guess that could get more complex. I could also imagine other complex things that could but probably shouldn't be done here.

It's not slow. It's making a new reference in the new env to the same pointer. 

> I guess I'm trying to get as intuitive (if you already know shiny) behaviour as possible here - could the output of the quosure actually be available in the module's environment? So that they behave as similar to reactives as possible.

Do you mean that objects from the quosure@env would be automatically or not available in the reactive?
We can still use `q[["object"]]` instead. I don't think mixing these two environments is a good idea.
1. Code could be successful if some object was in the reactive environment and not in quosure@env. 
2. We will need a way to pass `quosure@env` to the `reactive(..., env = )`
3. We don't want to include objects in the quosure@env which are not generated by the code. Often in the reactives there are objects which are not part of the code (proxy variables, observers)

> Do you actually want separate environments for everything - could you have a separate single "evaluation environment" for the shiny module? Would that be easier for module developers to grasp or harder?
I don't get it. Let's talk about this on thursday